src/nnbench/runner.py

```diff
@@ -236,16 +236,20 @@ def run(
 
         self._check(dparams)
 
-        ctx = Context()
+        if isinstance(context, Context):
+            ctx = context
+        else:
+            ctx = Context()
+
+            for provider in context:
+                ctx_cand = Context.make(provider())
 
-        for provider in context:
-            ctxval = provider()
-            # we do not allow multiple values for a context key.
-            duplicates = set(ctx.keys()) & set(ctxval.keys())
-            if duplicates:
-                dupe, *_ = duplicates
-                raise ValueError(f"got multiple values for context key {dupe!r}")
-            ctx.update(ctxval)
+                # we do not allow multiple values for a context key.
+                duplicates = set(ctx.keys()) & set(ctx_cand.keys())
+                if duplicates:
+                    dupe, *_ = duplicates
+                    raise ValueError(f"got multiple values for context key {dupe!r}")
+                ctx.update(ctx_cand)
```

Comment:
This method is suboptimal with the unique key requirement, but we could implement a key check under the hood in `Context.update()`, and add raw providers here in the loop with `Context.add()`. That's at most a follow-up, though.

Reply:
Nice, I can open an issue for that.